src/com/linkedin/parseq/Engine.java

```diff
@@ -113,25 +137,127 @@
       }
     };
 
-    if (_properties.containsKey(MAX_RELATIONSHIPS_PER_TRACE)) {
-      _maxRelationshipsPerTrace = (Integer) getProperty(MAX_RELATIONSHIPS_PER_TRACE);
-    } else {
-      _maxRelationshipsPerTrace = DEFUALT_MAX_RELATIONSHIPS_PER_TRACE;
-    }
   }
 
   public Object getProperty(String key) {
     return _properties.get(key);
   }
 
   /**
-   * Runs the given task with its own context. Use {@code Tasks.seq} and
-   * {@code Tasks.par} to create and run composite tasks.
+   * Runs the given task. Task passed in as a parameter becomes a root on a new Plan.
+   * All tasks created and started as a consequence of a root task will belong to that plan and will share a Trace.
+   * <p>
+   * This method blocks until Engine has a capacity to run the task. Engine's capacity is
+   * specified by a {@value #MAX_CONCURRENT_PLANS} configuration property. Use
+   * {@link EngineBuilder#setEngineProperty(String, Object)} to set this property.
+   * For the sake of backwards compatibility default value for a {@value #MAX_CONCURRENT_PLANS} is
+   * {@value #DEFUALT_MAX_CONCURRENT_PLANS} which essentially means "unbounded capacity".
    *
    * @param task the task to run
    */
   public void run(final Task<?> task) {
-    run(task, task.getClass().getName());
+    run(task, defaultPlanClass(task));
+  }
+
+  private final String defaultPlanClass(final Task<?> task) {
+    return task.getClass().getName();
+  }
+
+  /**
+   * Runs the given task. Task passed in as a parameter becomes a root on a new Plan.
+   * All tasks created and started as a consequence of a root task will belong to that plan and will share a Trace.
+   * <p>
+   * This method blocks until Engine has a capacity to run the task. Engine's capacity is
+   * specified by a {@value #MAX_CONCURRENT_PLANS} configuration property. Use
+   * {@link EngineBuilder#setEngineProperty(String, Object)} to set this property.
+   * For the sake of backwards compatibility default value for a {@value #MAX_CONCURRENT_PLANS} is
+   * {@value #DEFUALT_MAX_CONCURRENT_PLANS} which essentially means "unbounded capacity".
+   *
+   * @param task the task to run
+   * @param planClass string that identifies a "class" of the Plan. Plan class ends up in a ParSeq
+   * Trace and can be used to group traces into "classes" when traces are statistically analyzed.
+   */
+  public void run(final Task<?> task, final String planClass) {
+    try {
+      _concurrentPlans.acquire();
```

Comment:
Non-fair Semaphore.aquire() is implemented as an int CAS and falls back into locking only when number of permits is exhausted, so I don't think there is lock contention risk as long as Engine does not reach it's capacity limit. Once Engine reaches it's capacity limit we do expect locking and blocking.